src/EditorFeatures/Test2/Diagnostics/DiagnosticServiceTests.vb

```diff
@@ -2329,6 +2330,84 @@ class MyClass
+                               </Document>
+                           </Project>
+                       </Workspace>
+
+            Using workspace = TestWorkspace.CreateWorkspace(test, composition:=s_compositionWithMockDiagnosticUpdateSourceRegistrationService)
+                Dim solution = workspace.CurrentSolution
+                Dim project = solution.Projects.Single()
+
+                ' Add syntax and semantic analyzers
+                Dim syntaxAnalyzer = New FieldAnalyzer("ID0001", syntaxTreeAction:=True)
+                Dim semanticAnalyzer = New FieldAnalyzer("ID0002", syntaxTreeAction:=False)
+                Dim compilerAnalyzer = New CSharpCompilerDiagnosticAnalyzer()
+                Dim analyzerReference = New AnalyzerImageReference(ImmutableArray.Create(Of DiagnosticAnalyzer)(compilerAnalyzer, syntaxAnalyzer, semanticAnalyzer))
+                project = project.AddAnalyzerReference(analyzerReference)
+
+                Dim mefExportProvider = DirectCast(workspace.Services.HostServices, IMefHostExportProvider)
+                Assert.IsType(Of MockDiagnosticUpdateSourceRegistrationService)(workspace.GetService(Of IDiagnosticUpdateSourceRegistrationService)())
+                Dim diagnosticService = Assert.IsType(Of DiagnosticAnalyzerService)(workspace.GetService(Of IDiagnosticAnalyzerService)())
+                Dim incrementalAnalyzer = diagnosticService.CreateIncrementalAnalyzer(workspace)
+
+                ' Get diagnostics for span for the given DiagnosticKinds
+                Dim document = project.Documents.Single()
+                Dim root = Await document.GetSyntaxRootAsync()
+                Dim diagnostics = Await diagnosticService.GetDiagnosticsForSpanAsync(document, root.FullSpan, diagnosticKinds:=diagnosticKinds)
+
+                Dim expectedCount = 0
+                Dim expectedDiagnosticIds As New HashSet(Of String)
+
+                If (diagnosticKinds And DiagnosticKinds.CompilerSyntax) <> 0 Then
+                    expectedCount += 1
+                    expectedDiagnosticIds.Add("CS1513")
+                End If
+
+                If (diagnosticKinds And DiagnosticKinds.CompilerSemantic) <> 0 Then
+                    expectedCount += 1
+                    expectedDiagnosticIds.Add("CS0219")
+                End If
+
+                If (diagnosticKinds And DiagnosticKinds.AnalyzerSyntax) <> 0 Then
+                    expectedCount += 1
+                    expectedDiagnosticIds.Add("ID0001")
+                End If
+
+                If (diagnosticKinds And DiagnosticKinds.AnalyzerSemantic) <> 0 Then
+                    expectedCount += 1
+                    expectedDiagnosticIds.Add("ID0002")
+                End If
+
+                Assert.Equal(expectedCount, diagnostics.Length)
+                Dim actualDiagnosticIds = diagnostics.Select(Function(d) d.Id).ToHashSet()
+                Assert.Equal(expectedDiagnosticIds, actualDiagnosticIds)
+            End Using
+        End Function
```

Comment:
can we have some sort of test that shows that if you ask for semantics, then ask for syntax that you get the correct results in both cases (And vice versa)?

Reply:
Added